pinot-controller/src/main/java/org/apache/pinot/controller/cursors/ResponseStoreCleaner.java

```diff
@@ -0,0 +1,222 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.cursors;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import java.io.IOException;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.hc.client5.http.classic.methods.HttpDelete;
+import org.apache.hc.client5.http.classic.methods.HttpGet;
+import org.apache.hc.client5.http.classic.methods.HttpUriRequestBase;
+import org.apache.hc.client5.http.impl.io.PoolingHttpClientConnectionManager;
+import org.apache.hc.core5.http.io.entity.EntityUtils;
+import org.apache.helix.model.InstanceConfig;
+import org.apache.pinot.common.auth.AuthProviderUtils;
+import org.apache.pinot.common.http.MultiHttpRequest;
+import org.apache.pinot.common.http.MultiHttpRequestResponse;
+import org.apache.pinot.common.metrics.ControllerMetrics;
+import org.apache.pinot.common.response.CursorResponse;
+import org.apache.pinot.common.response.broker.CursorResponseNative;
+import org.apache.pinot.controller.ControllerConf;
+import org.apache.pinot.controller.LeadControllerManager;
+import org.apache.pinot.controller.api.resources.InstanceInfo;
+import org.apache.pinot.controller.helix.core.PinotHelixResourceManager;
+import org.apache.pinot.controller.helix.core.periodictask.ControllerPeriodicTask;
+import org.apache.pinot.spi.auth.AuthProvider;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.apache.pinot.spi.utils.JsonUtils;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * ResponseStoreCleaner periodically gets all responses stored in a response store and deletes the ones that have
+ * expired. From each broker, tt gets the list of responses. Each of the response has an expiration unix timestamp.
+ * If the current timestamp is greater, it calls a DELETE API for every response that has expired.
+ */
+public class ResponseStoreCleaner extends ControllerPeriodicTask<Void> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ResponseStoreCleaner.class);
+  private static final int TIMEOUT_MS = 3000;
+  private static final String QUERY_RESULT_STORE = "%s://%s:%d/responseStore";
+  private static final String DELETE_QUERY_RESULT = "%s://%s:%d/responseStore/%s";
+  // Used in tests to trigger the delete instead of waiting for the wall clock to move to an appropriate time.
+  public static final String CLEAN_AT_TIME = "response.store.cleaner.clean.at.ms";
```

Comment:
Can't we just set the frequency config to a much smaller number (few seconds maybe) instead of introducing this hidden configuration?

Reply:
No. I am forced to use `wait` because `scheduleNow` runs the test in another thread. So its an artifact of the functions availabe to me in the `PeriodicTask` class. As an overview, I want to test the following scenario:

```
At t1, run q1
At t2, run cleaner
At t3, run q2

where t1 < t2 < t3.

assert that only q1 is deleted
```

So instead of kicking off periodic scheduler at small periodicity value like 5s and then hoping to have t1 and t3 on either side of the 5s runs, I have forcefully serialized it by using the parameter.
What I unfortunately tripped on is that `scheduleNow` uses an executor service to schedule. So `scheduleNow` returns immediately and I have to wait for the executor thread to finish. However it is still deterministically serialized and therefore not flaky.

Reply:
One option to reduce confusion is to add `scheduleNowInline` or some such name which does not use an executor service.

Reply:
I see, I didn't take a close enough look at the tests to realize that you wanted to test a scenario where one result is deleted but not another.

Reply:
Assuming you were talking about `CursorIntegrationTest::testResponseStoreCleaner` here - doesn't that test only verify that at least one of the two response stores is deleted? In that case, why do we need this `response.store.cleaner.clean.at.ms` internal configuration? Wouldn't a short response store expiration value in the broker combined with a short frequency for the controller periodic job do the trick considering we're using a 100 second timeout for the wait condition anyway?

Reply:
Yes. That is the test I am referring to. Even this approach is flaky and the clue is in the wording of your question. What is the precise values for  `short response store expiration value` & `a short frequency for the controller periodic job` ? The fact that I have to go through multiple iterations to find these values is a test design smell. These values may differ based on whether it runs my laptop or a VM with a noisy neighbor.

Instead the description for my test is precise. I will run two queries 50 milliseconds apart. I will simulate a cleaner run to at the submission time of the first one. Since the check is LESS_THAN_EQUAL, only the first one has to be deleted.